File: octave/newton_interp.m
function Y = newton_interp(x, y, X)
    %
    % Newton interpolation method.
    %
    % Input:
    %   - x : precomputed points x axis values
    %   - y : precomputed points y axis values
    %   - X : query points x axis
    % Output:
    %   - Y : query points y axis (interpolation results)
    %
    if (size(x,2) != size(y,2))
        error('x and y must have the same number of elements')
    end
    n = size(x, 2); 
    Y = zeros(1,size(X, 2));
    for k = 1:n
        L = ones(1, size(X, 2));
        for i = 1:(k-1)
            L = L .* (X - x(i));
        end
        a = divdiff(x(1:k), y(1:k));
        Y = Y + a*L;
    end
end

File: python/newton_interp.py
# Newton interpolating polynomial

import numpy as np
from divdiff import divdiff

def newton_interp(x, y, X):
    '''
    Newton interpolation method
    Input:
      - x : precomputed points x axis values
      - y : precomputed points y axis values
      - X : query points x axis
    Output:
      - Y : query points y axis (interpolation results)
    '''
    n = len(x)
    Y = np.zeros(len(X))
    for k in range(n):
        L = np.ones(len(X))
        for i in range(k):
            L *= (X - x[i])
        a = divdiff(x[0:k+1], y[0:k+1])
        Y += a*L
    return Y
